### PR TITLE
CutPlugin class for cuts in straxen

### PR DESCRIPTION
**Pull request by jorana**
The cuts currently in straxen (https://github.com/XENONnT/straxen/blob/master/straxen/plugins/x1t_cuts.py ) have a quite lot of overlap with one another. Most of the time a simple Boolean array is all that is returned. However, we each time have to add the time fields et cetera which may somewhat obscure the functionality.
To simplify the cuts somewhat I've written a simple Plugin Class. In straxen it could be used as for example:
```
class CutS1s(CutPlugin):
    depends_on = ('events', 'corrected_areas')
    provides = 'cs1_cut'
    cut_description = "Remove cS1 above 100 PE"
    cut_name ='cut_high_cs1'
    __version__ = '0.0.0'
    
    def cut_by(self, events):
        return events['cs1'] > 100
```
```
class CutS2s(CutPlugin):
    depends_on = ('events', 'corrected_areas')
    cut_description = "Remove cS2s below 100 PE"
    
    def cut_by(self, events):
        return events['cs2'] < 100
```

```
class CutRecord0(CutPlugin):
    depends_on = ('records',)
    provides = 'is_not_record0_cut'
    cut_description = "is record 0"
    cut_name ='is_not_record0_cut'
    __version__ = '0.0.0'
    
    def cut_by(self, records):
        return records['record_i'] != 0
```

### CODE REVIEW

**Comment by WenzDaniel** [comment]
Hej Joran, 

nice Idea. I have not checked your code yet, but one quick question. In your current version cuts would be the same data_kind as events would not they be? Maybe it is better to have a new "cut"-data_kind instead?

**Comment by jorana** [comment]
> 
> 
> Hej Joran,
> 
> nice Idea. I have not checked your code yet, but one quick question. In your current version cuts would be the same data_kind as events would not they be? Maybe it is better to have a new "cut"-data_kind instead?

Hi Daniel,
Thanks for the suggestion! The datakind is derived from the dependency specified in the ```cut_by(self <datakind>)```, so in principle you could cut on peak level. I think it would make analysts lives harder if we create a new datatype that is not events as that is where most analysts will actually apply the cuts. That can get tricky if you want to load multiple datakinds as ```get_df``` (the function most analysts will use) will complain that this is not possible.

File: strax/plugin.py [hunk 1]
```diff
@@ -847,3 +847,50 @@ def cleanup(self, iters, wait_for):
...
+        else:
+            raise NotImplementedError(f"{self.cut_name} does not have attribute 'cut_by'")
+
+        # Take shape of the first data_type like in strax.plugin
+        buff = list(kwargs.values())[0]
+
+        # Generate result buffer
+        r = np.zeros(len(buff), self.dtype)
+        r['time'] = buff['time']
+        r['endtime'] = buff['endtime']
```
Comment (WenzDaniel) [inline strax/plugin.py:890]: This will only work if the data_kind has an endtime field. Which is not the case for the lower levels.

File: strax/plugin.py [hunk 2]
```diff
@@ -847,3 +847,50 @@ def cleanup(self, iters, wait_for):
...
+            else:
+                _description = " ".join(_description.split("_"))
+            self.cut_description = _description
+
+    def infer_dtype(self):
+        dtype = [(self.cut_name, np.bool, self.cut_description)]
+        dtype = dtype + strax.time_fields
+        return dtype
+
+    def compute(self, **kwargs):
```
Comment (WenzDaniel) [inline strax/plugin.py:878]: To get the start and endtime of your data you can do 

`compute(<data_kinds>, start, end)` 

See: https://github.com/AxFoundation/strax/blob/7c0641000301653adbf4144967b8a9cc18bf8103/strax/plugin.py#L448 and my comment below.

File: strax/plugin.py [hunk 3]
```diff
@@ -847,3 +847,50 @@ def cleanup(self, iters, wait_for):
...
+        if not hasattr(self, 'cut_description'):
+            _description = _name
+            if 'cut_' not in _description:
+                _description = 'Cut by ' + _description
+            else:
+                _description = " ".join(_description.split("_"))
+            self.cut_description = _description
+
+    def infer_dtype(self):
+        dtype = [(self.cut_name, np.bool, self.cut_description)]
```
Comment (WenzDaniel) [inline strax/plugin.py:874]: Does not it have to be np.bool_ ? https://numpy.org/doc/stable/user/basics.types.html

File: strax/plugin.py (hunk 1 above)
Comment (jorana) [reply strax/plugin.py:890]: thanks very good point!

File: strax/plugin.py (hunk 2 above)
Comment (jorana) [reply strax/plugin.py:878]: I will use strax.endtime as it is somewhat easier

File: strax/plugin.py (hunk 3 above)
Comment (jorana) [reply strax/plugin.py:874]: thanks, didn't know that, will change it!

**Comment by jorana** [comment]
As @WenzDaniel just correctly pointed out the endtime needs to be computed when we are working with lower level plugins. Although it's not really the envisaged application it's easy enough to implement.

See now for example:
```
class CutRecord0(CutPlugin):
    depends_on = ('records',)
    provides = 'is_not_record0_cut'
    cut_description = "is record 0"
    cut_name ='is_not_record0_cut'
    __version__ = '0.0.0'
    
    def cut_by(self, records):
        return records['record_i'] != 0
```

**Comment by jorana** [comment]
In 752f521 I've added a test mainly since I think I should learn how how to write tests. Also it's completely fine to add more tests.

File: strax/plugin.py [hunk 4]
```diff
@@ -847,3 +847,51 @@ def cleanup(self, iters, wait_for):
...
+            cut_by = self.cut_by
+        else:
+            raise NotImplementedError(f"{self.cut_name} does not have attribute 'cut_by'")
+
+        # Take shape of the first data_type like in strax.plugin
+        buff = list(kwargs.values())[0]
+
+        # Generate result buffer
+        r = np.zeros(len(buff), self.dtype)
+        r['time'] = buff['time']
```
Comment (WenzDaniel) [inline strax/plugin.py:890]: You have to add a test whether buff will be an empty. In that case buff['time'] will crash.

Comment (jorana) [reply strax/plugin.py:890]: Actually, you can make copies of empty data. For example the line below does not fail:
```
strax.endtime(np.zeros(0, dtype=strax.time_fields))
```

**Comment by jorana** [comment]
@ ACs this is ready for merge

File: tests/test_cut_plugin.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    st = strax.Context(storage=[])
+    st.register(ToBeCut)
+    st.register(CutSomething)
+
+    result = st.get_array(run_id='some_run',
+                          targets=strax.camel_to_snake(CutSomething.__name__))
+    correct_answer = np.sum(input_peaks[_dtype_name] > cut_threshold)
+    assert len(result) == len(input_peaks), "WTF??"
+    assert correct_answer == np.sum(result['cut_something']), "Cut plugin does not give boolean arrays correctly"
```
Comment (WenzDaniel) [inline tests/test_cut_plugin.py:100]: Maybe just add one additional test checking the endtime since we had these issues in peaks and you are computing some new endtimes here. Does not hurt much to have this little check. All the rest looks fine.

Comment (jorana) [reply tests/test_cut_plugin.py]: Thanks, this is done in 3673c28 !
